### DISCUSSION TITLE
Configuring credentials/tokens/environment variables

### DISCUSSION THREAD

**Discussion by cobblermck** [renovatebot/renovate#21052] [Request Help]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

3.9

### If you're self-hosting Renovate, select which platform you are using.

Bitbucket Server

### If you're self-hosting Renovate, tell us what version of the platform you run.

7.6.17

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

Using docker compose in a simple Jenkins pipeline to create a Renovate instance.

### What you tried so far.

I tried running the Docker image instead of compose but we're using Docker Cloud in Jenkins and Docker inside of Docker didn't work quickly/easily so I moved back to docker compose, which was the original intent

I have tried setting RENOVATE_TOKEN, GITHUB_COM_TOKEN, and RENOVATE_HOST_RULES environment variables and running docker compose on my local machine as well as in our Jenkins pipeline to no avail. I used to have the Bitbucket username in the config.js and a blank password, hoping to pass in the env for all secrets (the three listed above). I removed the password line entirely, and then the username entirely as well, from config.js.

I don't have any trouble running it locally with all of the secrets added to the config.js and docker-compile.yml files. I have not tried running it in Jenkins with them but I assume it would work as well. It's just the secrets management that is throwing me for a loop.

### Relevant debug logs

<details><summary>Logs</summary>

```
renovate-renovate-1  | DEBUG: Converting GITHUB_COM_TOKEN into a global host rule
renovate-renovate-1  | DEBUG: File config
renovate-renovate-1  |        "config": {
renovate-renovate-1  |          "endpoint": "<redact>",
renovate-renovate-1  |          "platform": "bitbucket-server",
renovate-renovate-1  |          "onboardingConfig": {"extends": ["config:base", "group:all"]},
renovate-renovate-1  |          "repositories": ["<redact>"],
renovate-renovate-1  |          "baseBranches": ["development"],
renovate-renovate-1  |          "commitMessagePrefix": "<redact>",
renovate-renovate-1  |          "detectHostRulesFromEnv": true,
renovate-renovate-1  |          "packageRules": [
renovate-renovate-1  |            {"matchUpdateTypes": ["patch"], "enabled": false},
renovate-renovate-1  |            {"matchPackagePatterns": ["<redact>:*"], "enabled": false}
renovate-renovate-1  |          ]
renovate-renovate-1  |        }
renovate-renovate-1  | DEBUG: CLI config
renovate-renovate-1  |        "config": {}
renovate-renovate-1  | DEBUG: Env config
renovate-renovate-1  |        "config": {
renovate-renovate-1  |          "hostRules": [
renovate-renovate-1  |            {"hostType": "github", "matchHost": "github.com", "token": "***********"}
renovate-renovate-1  |          ]
renovate-renovate-1  |        }
renovate-renovate-1  | DEBUG: Combined config
renovate-renovate-1  |        "config": {
renovate-renovate-1  |          "endpoint": "<redact>",
renovate-renovate-1  |          "platform": "bitbucket-server",
renovate-renovate-1  |          "onboardingConfig": {"extends": ["config:base", "group:all"]},
renovate-renovate-1  |          "repositories": ["<redact>"],
renovate-renovate-1  |          "baseBranches": ["development"],
renovate-renovate-1  |          "commitMessagePrefix": "<redact>",
renovate-renovate-1  |          "detectHostRulesFromEnv": true,
renovate-renovate-1  |          "packageRules": [
renovate-renovate-1  |            {"matchUpdateTypes": ["patch"], "enabled": false},
renovate-renovate-1  |            {"matchPackagePatterns": ["<redact>:*"], "enabled": false}
renovate-renovate-1  |          ],
renovate-renovate-1  |          "hostRules": [
renovate-renovate-1  |            {"hostType": "github", "matchHost": "github.com", "token": "***********"}
renovate-renovate-1  |          ]
renovate-renovate-1  |        }
renovate-renovate-1  | DEBUG: Adding trailing slash to endpoint
renovate-renovate-1  | DEBUG: Enabling forkProcessing while in non-autodiscover mode
renovate-renovate-1  | DEBUG: Found valid git version: 2.40.0
renovate-renovate-1  | FATAL: You must configure a Bitbucket Server username/password
renovate-renovate-1  | DEBUG: Renovate exiting
renovate-renovate-1  |  INFO: Renovate is exiting with a non-zero code due to the following logged errors
renovate-renovate-1  |        "loggerErrors": [
renovate-renovate-1  |          {
renovate-renovate-1  |            "name": "renovate",
renovate-renovate-1  |            "level": 60,
renovate-renovate-1  |            "logContext": "SHcUsYRhmFZEJt7h0NX8m",
renovate-renovate-1  |            "msg": "You must configure a Bitbucket Server username/password"
renovate-renovate-1  |          }
renovate-renovate-1  |        ]
renovate-renovate-1 exited with code 1

```

</details>

**Comment by rarkins** [comment]
The logs you posted indicate GITHUB_COM_TOKEN has been detected in env but the others have not (notably RENOVATE_TOKEN). This indicates that env is being passed but something is wrong with the particular variable. Somehow you have a local/environment problem so it's hard to troubleshoot from afar

**Reply by cobblermck** [reply]
They're all set the same way - I just did an 'export ENV_VARIABLE=value' for all three. If I 'echo' I can see that their values are set. But you're right, there was something different about how I was handling the Github token vs. the others - I had it defined in the environment section of my docker-compose.yml thusly:
`GITHUB_COM_TOKEN=process.env.GITHUB_COM_TOKEN`
So I added the others:
`RENOVATE_TOKEN=process.env.RENOVATE_TOKEN`
`RENOVATE_HOST_RULES=process.env.RENOVATE_HOST_RULES`

Now my error is:
`renovate-renovate-1  | DEBUG: Could not parse environment variable
renovate-renovate-1  |        "val": "process.env.RENOVATE_HOST_RULES",
renovate-renovate-1  |        "envName": "RENOVATE_HOST_RULES"
`

`echo $RENOVATE_HOST_RULES
[{"matchHost":"artifactoryURL","username":"artifactoryUsername","password":"artifactoryAPIKey"}]`

**Reply by cobblermck** [reply]
Turned on my brain and put the hostRules back in my config.js like this:
`hostRules: [
        {
            matchHost: 'artifactoryURL',
            username: 'artifactoryUser',
            password: process.env.ARTIFACTORY_TOKEN
        }
    ],`

Now I don't get the parsing error, but it's still saying I haven't provided Bitbucket credentials
`renovate-renovate-1  | FATAL: You must configure a Bitbucket Server username/password`

`$ echo $RENOVATE_TOKEN
<my-token>`

**Reply by cobblermck** [reply]
Turned on my brain all the way and added username and password back to config.js, using process.env.RENOVATE_TOKEN as the value for password. It now gives me a 401 unauthorized error which I'm presuming to mean it's reading the value literally rather than taking the environment variable, but I only think that because I'm sure of the accuracy of the environment value - it is a specific token I just created for this purpose. Is this current configuration correct by your estimation?

**Reply by rarkins** [reply]
I'm pretty sure there is no concept of process.env in YAML files

**Reply by rarkins** [reply]
https://docs.docker.com/compose/environment-variables/set-environment-variables/

**Reply by cobblermck** [reply]
It doesn't seem to be having any trouble with the github token - at least until after I can get past the Bitbucket authentication which I can't seem to do right now. That token and my Artifactory token are both defined in my config.js rather than my docker-compose.yml, but you're correct, all three are currently defined by process.env at the moment

**Reply by cobblermck** [reply]
Here's where the sweet spot seems to be:
config.js has username set and password set to process.env.RENOVATE_TOKEN, hostRules has the username set and password set to process.env.ARTIFACTORY_TOKEN
docker-compose.yml has the following in the environment section: RENOVATE_TOKEN=process.env and GITHUB_COM_TOKEN=process.env

Then I set my environment variables
Then I run docker compose create renovate, docker compose run -e RENOVATE_TOKEN renovate

I got a nice clean run by doing this.

**Reply by viceice** [reply]
please provide the config.js and the docker-compose.yml
